File: typescript-client/s3Types.ts
/**
 * S3 object representation, either as a URI string or a record object
 */
export type S3Object = S3ObjectURI | S3ObjectRecord;

/**
 * S3 object URI in the format `s3://storage/key` (`s3:///key` targets the
 * workspace default storage)
 */
export type S3ObjectURI = `s3://${string}/${string}`;

/**
 * S3 object record with file key, optional storage identifier, and optional presigned token
 */
export type S3ObjectRecord = {
  /** File key/path in S3 bucket */
  s3: string;
  /** Storage backend identifier */
  storage?: string;
  /** Presigned URL query string for public access */
  presigned?: string;
};

/**
 * S3 client configuration settings for Deno S3 light client
 */
export type DenoS3LightClientSettings = {
  /** S3 endpoint URL */
  endPoint: string;
  /** AWS region */
  region: string;
  /** Bucket name */
  bucket?: string;
  /** Use HTTPS connection */
  useSSL?: boolean;
  /** AWS access key */
  accessKey?: string;
  /** AWS secret key */
  secretKey?: string;
  /** Use path-style URLs instead of virtual-hosted style */
  pathStyle?: boolean;
};

/**
 * Parse an S3 object from URI string or record format
 * @param s3Object - S3 object as URI string (`s3://storage/key`, `s3:///key`
 *   for the default storage) or record. Any other string throws rather than
 *   falling back to an auto-generated key: an auto key is requested by
 *   omitting the object, and a fallback would silently misplace the upload
 *   on any typo.
 * @returns S3 object record with storage and s3 key
 */
export function parseS3Object(s3Object: S3Object): S3ObjectRecord {
  if (typeof s3Object === "object") return s3Object;
  const match = s3Object.match(/^s3:\/\/([^/]*)\/(.+)$/);
  if (match) return { storage: match[1] || undefined, s3: match[2] };
  if (s3Object.startsWith("s3://")) {
    throw new Error(
      `Invalid s3 object URI ${JSON.stringify(s3Object)}: expected s3://<storage>/<key> with a non-empty key (s3:///<key> for the default storage)`
    );
  }
  throw new Error(
    `Invalid s3 object ${JSON.stringify(s3Object)}: expected an s3://<storage>/<key> URI (e.g. "s3:///${s3Object}" for key "${s3Object}" in the default storage) or { s3: <key> }`
  );
}

/**
 * Settings necessary to connect DuckDB to an S3 bucket
 */
export type DuckDbConnectionSettings = {
  /** DuckDB SET statements to configure the S3 connection */
  connection_settings_str: string;
  /** Azure container path when the target storage is Azure */
  azure_container_path?: string;
};

/**
 * S3 filesystem arguments consumed by Python's Polars `scan_parquet`/`read_parquet`
 */
export type PolarsS3FsArgs = {
  endpoint_url: string;
  key?: string;
  secret?: string;
  use_ssl: boolean;
  cache_regions?: boolean;
  client_kwargs?: Record<string, unknown>;
};

/**
 * Object-storage options consumed by Python's Polars when reading S3 data
 */
export type PolarsStorageOptions = {
  aws_endpoint_url: string;
  aws_region: string;
  aws_allow_http: string;
  aws_access_key_id?: string;
  aws_secret_access_key?: string;
};

/**
 * Settings necessary to connect Polars to an S3 bucket
 */
export type PolarsConnectionSettings = {
  s3fs_args: PolarsS3FsArgs;
  storage_options?: PolarsStorageOptions;
};

/**
 * Settings necessary to connect a boto3 client to an S3 bucket
 */
export type Boto3ConnectionSettings = {
  endpoint_url: string;
  region_name: string;
  use_ssl: boolean;
  aws_access_key_id: string;
  aws_secret_access_key: string;
  aws_session_token?: string;
};

File: python-client/wmill/wmill/s3_types.py
from typing import Optional


class S3Object(dict):
    """S3 file reference with file key, optional storage identifier, and presigned token."""
    s3: str
    storage: Optional[str]
    presigned: Optional[str]

    def __getattr__(self, attr):
        return self[attr]


class S3FsClientKwargs(dict):
    """S3FS client keyword arguments for region configuration."""
    region_name: str

    def __getattr__(self, attr):
        return self[attr]


class S3FsArgs(dict):
    """S3FS connection arguments including endpoint, credentials, and client settings."""
    endpoint_url: str
    key: str
    secret: str
    use_ssl: bool
    cache_regions: bool
    client_kwargs: S3FsClientKwargs

    def __getattr__(self, attr):
        return self[attr]


class StorageOptions(dict):
    """Storage options for Polars S3 connectivity with AWS credentials and endpoint."""
    aws_endpoint_url: str
    aws_access_key_id: str
    aws_secret_access_key: str
    aws_region: str
    aws_allow_http: str

    def __getattr__(self, attr):
        return self[attr]


class PolarsConnectionSettings(dict):
    """Polars S3 connection settings containing S3FS args and storage options."""
    s3fs_args: S3FsArgs
    storage_options: StorageOptions

    def __getattr__(self, attr):
        return self[attr]


class Boto3ConnectionSettings(dict):
    """Boto3 S3 connection settings with endpoint, region, and AWS credentials."""
    endpoint_url: str
    region_name: str
    use_ssl: bool
    aws_access_key_id: str
    aws_secret_access_key: str

    def __getattr__(self, attr):
        return self[attr]


class DuckDbConnectionSettings(dict):
    """DuckDB S3 connection settings as a configuration string."""
    connection_settings_str: str

    def __getattr__(self, attr):
        return self[attr]
